crates/sui-indexer-alt-graphql/src/api/types/epoch.rs

```diff
@@ -240,28 +256,28 @@ impl EpochStart {
 }
 
 impl EpochEnd {
-    fn empty() -> Self {
-        Self { contents: None }
-    }
-
     /// Attempt to fetch information about the end of an epoch from the store. May return an empty
     /// response if the epoch has not ended yet, as of the checkpoint being viewed.
-    async fn fetch(ctx: &Context<'_>, scope: &Scope, epoch_id: u64) -> Result<Self, RpcError> {
+    async fn fetch(
+        ctx: &Context<'_>,
+        epoch_start: &EpochStart,
+        epoch_id: u64,
+    ) -> Result<Self, RpcError> {
         let pg_loader: &Arc<DataLoader<PgReader>> = ctx.data()?;
-        let Some(stored) = pg_loader
+        let scope = epoch_start.scope.clone();
+        let stored = pg_loader
             .load_one(EpochEndKey(epoch_id))
             .await
-            .context("Failed to fetch epoch end information")?
-        else {
-            return Ok(Self::empty());
-        };
-
-        if stored.cp_hi as u64 > scope.checkpoint_viewed_at() {
-            return Ok(Self::empty());
-        }
+            .context("Failed to fetch epoch end information")?;
 
+        let end = match stored {
+            Some(end) if end.cp_hi as u64 <= scope.checkpoint_viewed_at() => Some(Arc::new(end)),
```

Comment:
I think because of the inclusive vs exclusive bounds here, we might be off by one on this test? If `scope.checkpoint_viewed_at()` was equal to `end.cp_hi - 1` then we should still treat the epoch as having ended (you can use `Checkpoint.query` to construct a test for this).

You will need to be careful about not introducing an underflow or an overflow when doing this though (not subtracting `1` from `0` or adding `1` to a value that could be `u64::MAX` -- which may happen if the value is user input). I don't think you need to worry too hard about the overflow case here, because even though `scope.checkpoint_viewed_at` can come from the user, we check that it corresponds to a valid checkpoint, which means it's not going to reach `u64::MAX` any time soon.